auth_proxy/services.py

```diff
@@ -53,6 +54,25 @@ def audit(self, client_id):
         return self.db.session.query(Token).\
             filter_by(client_id=client_id)
 
+    def smart_token_credentials(self, grant_type, code=None, refresh_token=None):
+        """ Provide additional credentials required by a SMART token request.
+        """
+        if grant_type == 'authorization_code':
+            grant = self.db.session.query(Grant).\
+                filter_by(code=code).first()
+            user = grant.user
```

Comment:
So, line 93 isn't quite right because that's the time at which the grant (authorization code) is created. The check I'm talking about needs to happen at the time when the code is exchanged for a token (basically asking: is the session that was active when the grant was created the same as the session active now?) Flask-oauthlib doesn't appear to have this protection built-in.